util/conditions/setter.go

```diff
@@ -0,0 +1,147 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package conditions
+
+import (
+	"fmt"
+	"sort"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+)
+
+// Setter interface defines methods that a Cluster API object should implement in order to
+// use the conditions package for setting conditions.
+type Setter interface {
+	Getter
+	SetConditions(clusterv1.Conditions)
+}
+
+// Set sets the given condition.
+// Please note that a condition is set only when a state transition happens, where a state transition
+// is define by a change of one of the following field: Status, Reason, Severity and Message.
+// During a state transition, the LastTransitionTime is automatically set.
+func Set(to Setter, condition *clusterv1.Condition) {
+	if to == nil || condition == nil {
+		return
+	}
+
+	conditions := to.GetConditions()
+	newConditions := make(clusterv1.Conditions, 0, len(conditions))
+	exists := false
+	condition.LastTransitionTime = metav1.Now()
+	for i := range conditions {
+		existingCondition := conditions[i]
+		if existingCondition.Type == condition.Type {
+			exists = true
+			if !hasSameState(&existingCondition, condition) {
+				newConditions = append(newConditions, *condition)
+				continue
+			}
+		}
+		newConditions = append(newConditions, existingCondition)
+	}
+	if !exists {
+		newConditions = append(newConditions, *condition)
+	}
+
+	// Sorts conditions for convenience of the consumer, i.e. kubectl.
+	sort.Slice(newConditions, func(i, j int) bool {
+		return lexicographicLess(&newConditions[i], &newConditions[j])
+	})
```

Comment:
Can we make `clusterv1.Conditions` to fulfill the sorting interface?

Reply:
I'm trying to avoid implementing code in the versioned API, and in this case, sorting should be used only in this package.